### PR TITLE
Initial Review

### PR DESCRIPTION
**Pull request by Ranjeet-8**

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#include <iostream>
```
Comment (akashshettyjs) [inline main.cpp]: Use bits/stdc++.h for all required cpp headers in all files

File: main.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#include <vector>
+#include "FileRead.h"
+#include "Display.h"
+
+using namespace std;
+
+int main(){
```
Comment (akashshettyjs) [inline main.cpp]:
```suggestion
int main() {
```

File: main.cpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    string file_name;
+    cout<<"Enter the File Name: "<<endl;
+    cin>>file_name;
+    int flag;
+    while(1){
```
Comment (akashshettyjs) [inline main.cpp]:
```suggestion

    while(1) {
```

File: main.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        
+        cout<<"--------------------------------------"<<endl;
+        cout<<"|         Filters Available          |"<<endl;
+        cout<<"|         1. Pid and Vid (Flag = 1)  |"<<endl;
+        cout<<"|         2. Time Range (Flag = 2)   |"<<endl;
+        cout<<"--------------------------------------"<<endl;
+        cin>>flag;
```
Comment (akashshettyjs) [inline main.cpp:19]:
```suggestion

        cin>>flag;
```

File: main.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+        FileRead file_object(file_name);
+        file_object.readFile();
+        vector<LogLine> dataset = file_object.getDataset();
+
+        if(flag==1){
```
Comment (akashshettyjs) [inline main.cpp]:
```suggestion
        if(flag == 1) {
```

File: main.cpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            string pid,vid;
```
Comment (akashshettyjs) [inline main.cpp]:
```suggestion
            string pid, vid;
```

Comment (akashshettyjs) [reply main.cpp]: maintain single space between every literals

File: main.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            cout<<"Enter the value of PID and VID: "<<endl;
+            cin>>pid>>vid;
+            Display pobj(dataset);
+            pobj.printPV(pid,vid);
```
Comment (akashshettyjs) [inline main.cpp]:
```suggestion
            pobj.printPV(pid, vid);
```

File: main.cpp [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        }
+        else if(flag==2){
```
Comment (akashshettyjs) [inline main.cpp]:
```suggestion
        else if(flag == 2) {
```
maintain single space between literals throughout the code

File: main.cpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            cout<<"Enter the value for start time and end time: "<<endl;
+            cin>>startTime>>endTime;
+            Display pobj(dataset);
+            pobj.printTimeRange(startTime,endTime);
+        }
+        else{
+            cout<<"Enter valid flag value as mentioned"<<endl;
+        }
+
+        return 0;
```
Comment (akashshettyjs) [inline main.cpp]: why is it returning after a single loop?

File: main.cpp [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
```
Comment (akashshettyjs) [inline main.cpp]: remove blank spaces

File: FileRead.h [hunk 11]
```diff
@@ -0,0 +1,27 @@
+#ifndef FILEREAD_H
+#define FILEREAD_H
+
+#include <iostream>
```
Comment (akashshettyjs) [inline FileRead.h]: bits/stdc++.h

File: FileRead.h [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+
+public:
+    FileRead(string input);
+
+    void readFile();
+
+    vector<LogLine> getDataset();
+};
+
+
```
Comment (akashshettyjs) [inline FileRead.h]: remove empty spaces

File: LogLine.h [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    string getDate();
+    map<string, int> getTime();
+    string getProcessId();
+    string getThreadId();
+    string getLogLevel();
+    string getLogMessage();
+};
+
+
+
```
Comment (akashshettyjs) [inline LogLine.h]: spaces

File: LogLine.cpp [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+LogLine::LogLine(string line) {
+    doParsing(line);
+}
+
+void LogLine::doParsing(string input) {
+    int set_count = 0, startIndex = 0;
+    string word;
+    for (int i = 0; i < input.length(); i++) {
+        if (input[i] == ' ') {
```
Comment (akashshettyjs) [inline LogLine.cpp:20]: add a logic here to consider multiple spaces as well

File: LogLine.cpp [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+            word = input.substr(startIndex, i - startIndex);
+            startIndex = i + 1;
+            set_count++;
+        }
+    }
+    setMemberVariableValues(word, set_count);
```
Comment (akashshettyjs) [inline LogLine.cpp]: member variable has to be set everytime set_count is changed, otherwise this logic wont work here

File: TimeParse.cpp [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+TimeParse::TimeParse(string line)
+{
+    doParsing(line);
+}
+void TimeParse::doParsing(string input)
+{
+    int set_count =0, startIndex =0;
+    string word;
+    for(int i=0;i<input.length();i++)
```
Comment (akashshettyjs) [inline TimeParse.cpp:11]: spacing

File: TimeParse.cpp [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        if(input[i] ==':'){
+            word = input.substr(startIndex,i-1);
```
Comment (akashshettyjs) [inline TimeParse.cpp]: How i - 1 ?, this is supposed to be i - startIndex ?

File: TimeParse.cpp [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+            startIndex=i+1;
+            set_count++;  
+        }
+            setMemberVariableValues(word,set_count);
```
Comment (akashshettyjs) [inline TimeParse.cpp:23]: member variable to be set only when set_count is changed, and not to set throughout for loop

File: TimeParse.cpp [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+    }
+}
+void TimeParse::setMemberVariableValues(string input, int sc)
+{
+    switch (sc)
+    {
+        case 1 :
```
Comment (akashshettyjs) [inline TimeParse.cpp]: use curly braces for switch cases

File: TimeParse.cpp [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+            cout<<"Class TimeParse : Function setMemberVariableValues : Getting into default section of switch case : check sc variable value "<<endl;
+            break;
+    }
+}
+int TimeParse::stringToInt(string input)
+{
+    int output = 0;
+    for(int i=0;i<input.length();i++)
+    {
+        output = output * 10 + input[i];
```
Comment (akashshettyjs) [inline TimeParse.cpp]: ASCII value of 0-9 is 48-57. This has to be considered while making this conversion here

Comment (akashshettyjs) [reply TimeParse.cpp]: also use parenthesis while having multiple calculations in same line

File: LogLine.cpp [hunk 21]
```diff
@@ -0,0 +1,84 @@
...
+LogLine::LogLine(string line) {
+    doParsing(line);
+}
+
+void LogLine::doParsing(string input) {
+    int set_count = 0, startIndex = 0;
+    string word;
+    for (int i = 0; i < input.length(); i++) {
+        if (input[i] == ' ') {
+            word = input.substr(startIndex, i - startIndex);
```
Comment (akashshettyjs) [inline LogLine.cpp:21]: add a logic to consider whole remaining line for LOG Message

File: Display.cpp [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+
+using namespace std;
+
+Display::Display(vector<LogLine>& set) : dataset(set) {}
+
+void Display::printPV(string pid, string vid)
+{
+    for(auto index : dataset)
+    {
+        if(index.getProcessId() == pid && index.getThreadId() == vid)
```
Comment (akashshettyjs) [inline Display.cpp]: use parenthesis to diff between two conditions

File: Display.cpp [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+    }
+}
+void Display::printTimeRange(string start, string end)
+{
+    for(auto index : dataset)
+    {
+        doParsing(start);
+        map<string,int> uTimemap;
+        uTimemap = index.getTime();
+        if(uTimemap.count("Hour") && uTimemap.count("Minute") && uTimemap.count("Second") && uTimemap.count("Millisecond"))
```
Comment (akashshettyjs) [inline Display.cpp]: it should be .contains() and not .count()

File: Display.cpp [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            if(uTimemap["Hour"]>=hour && uTimemap["Minute"]>=minute && uTimemap["Second"]>=second && uTimemap["Millisecond"]>=millisecond)
```
Comment (akashshettyjs) [inline Display.cpp]: use parenthesis between conditions

File: Display.cpp [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+
+    cout<<lobj.getDate()<<" "<<time["Hour"]<<":"<<time["Minute"]<<":"<<time["Second"]<<"."<<time["Millisecond"]<<" "<<lobj.getProcessId()<<" "<<lobj.getThreadId()<<" "<<lobj.getLogLevel()<<" "<<lobj.getLogMessage()<<endl;
+}
+void Display::doParsing(string input)
+{
+    int set_count =0, startIndex =0;
+    string word;
+    for(int i=0;i<input.length();i++)
+    {
+        if(input[i] ==':'){
```
Comment (akashshettyjs) [inline Display.cpp]: How are you parsing millisecs using this logic? millisec wont end with ' : '

File: Display.cpp [hunk 26]
```diff
@@ -0,0 +1,87 @@
...
+            word = input.substr(startIndex,i-1);
```
Comment (akashshettyjs) [inline Display.cpp]: not i-1

File: Display.cpp [hunk 27]
```diff
@@ -0,0 +1,87 @@
...
+            startIndex=i+1;
+            set_count++;  
+        }
+        setMemberVariableValues(word,set_count);
```
Comment (akashshettyjs) [inline Display.cpp]: setMemberVariableValues() not to be called throughout for loop

File: Display.cpp [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+    }
+}
+void Display::setMemberVariableValues(string input, int sc)
+{
+    switch (sc)
+    {
+        case 1 :
```
Comment (akashshettyjs) [inline Display.cpp]: curly braces

File: Display.cpp [hunk 29]
```diff
@@ -0,0 +1,87 @@
...
+            cout<<"Class Display : Function setMemberVariableValues : Getting into default section of switch case : check sc variable value "<<endl;
+            break;
+    }
+}
+int Display::stringToInt(string input)
+{
+    int output = 0;
+    for(int i=0;i<input.length();i++)
+    {
+        output = output * 10 + input[i];
```
Comment (akashshettyjs) [inline Display.cpp]: ASCII to be considered

**Review by akashshettyjs — COMMENTED** [review]
Please check all comments

File: main.cpp (hunk 4 above)
Comment (Ranjeet-8) [reply main.cpp:19]: Whats need to done ?

File: main.cpp (hunk 1 above)
Comment (Ranjeet-8) [reply main.cpp]: done

File: main.cpp (hunk 2 above)
Comment (Ranjeet-8) [reply main.cpp]: done

File: main.cpp (hunk 3 above)
Comment (Ranjeet-8) [reply main.cpp]: done

File: main.cpp (hunk 5 above)
Comment (Ranjeet-8) [reply main.cpp]: done

File: main.cpp (hunk 6 above)
Comment (Ranjeet-8) [reply main.cpp]: done

File: main.cpp (hunk 7 above)
Comment (Ranjeet-8) [reply main.cpp]: done

File: main.cpp (hunk 8 above)
Comment (Ranjeet-8) [reply main.cpp]: done

File: main.cpp (hunk 9 above)
Comment (Ranjeet-8) [reply main.cpp]: done

File: main.cpp (hunk 10 above)
Comment (Ranjeet-8) [reply main.cpp]: done

File: FileRead.h (hunk 11 above)
Comment (Ranjeet-8) [reply FileRead.h]: done

File: FileRead.h (hunk 12 above)
Comment (Ranjeet-8) [reply FileRead.h]: done

File: LogLine.h (hunk 13 above)
Comment (Ranjeet-8) [reply LogLine.h]: done

File: LogLine.cpp (hunk 14 above)
Comment (Ranjeet-8) [reply LogLine.cpp:20]: done

File: LogLine.cpp (hunk 15 above)
Comment (Ranjeet-8) [reply LogLine.cpp]: done

File: TimeParse.cpp (hunk 16 above)
Comment (Ranjeet-8) [reply TimeParse.cpp:11]: done

File: TimeParse.cpp (hunk 17 above)
Comment (Ranjeet-8) [reply TimeParse.cpp]: done

File: TimeParse.cpp (hunk 18 above)
Comment (Ranjeet-8) [reply TimeParse.cpp:23]: done

File: TimeParse.cpp (hunk 19 above)
Comment (Ranjeet-8) [reply TimeParse.cpp]: done

File: TimeParse.cpp (hunk 20 above)
Comment (Ranjeet-8) [reply TimeParse.cpp]: done

File: LogLine.cpp (hunk 21 above)
Comment (Ranjeet-8) [reply LogLine.cpp:21]: done

File: Display.cpp (hunk 22 above)
Comment (Ranjeet-8) [reply Display.cpp]: done

File: Display.cpp (hunk 23 above)
Comment (Ranjeet-8) [reply Display.cpp]: done.
introduced in c++20
may be not supported in cpp installed in my pc

File: Display.cpp (hunk 24 above)
Comment (Ranjeet-8) [reply Display.cpp]: done

File: Display.cpp (hunk 25 above)
Comment (Ranjeet-8) [reply Display.cpp]: done

File: Display.cpp (hunk 26 above)
Comment (Ranjeet-8) [reply Display.cpp]: done

File: Display.cpp (hunk 27 above)
Comment (Ranjeet-8) [reply Display.cpp]: done

File: Display.cpp (hunk 28 above)
Comment (Ranjeet-8) [reply Display.cpp]: dene

File: Display.cpp (hunk 29 above)
Comment (Ranjeet-8) [reply Display.cpp]: done

**Comment by akashshettyjs** [comment]
Push latest changes to git

File: main.cpp (hunk 4 above)
Comment (akashshettyjs) [reply main.cpp:19]: Line space after cin befor while

File: Display.cpp [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+        doParsing(start);
+        map<string,int> uTimemap;
+        uTimemap = index.getTime();
+        if((uTimemap.contains("Hour")) && (uTimemap.contains("Minute")) && (uTimemap.contains("Second")) && (uTimemap.contains("Millisecond")))
+        {
+            if((uTimemap["Hour"]>=hour) && (uTimemap["Minute"]>=minute) && (uTimemap["Second"]>=second) && (uTimemap["Millisecond"]>=millisecond))
+            {
+                doParsing(end);
+            }
+            if((uTimemap["Hour"]<=hour) && (uTimemap["Minute"]<=minute) && (uTimemap["Second"]<=second) && (uTimemap["Millisecond"]<=millisecond))
```
Comment (akashshettyjs) [inline Display.cpp]: This if condition has to be inside previous if { }

File: TimeParse.cpp [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        if(set_count == 3) {
+            word = input.substr(startIndex,(input.length()-1)-startIndex);
+            set_count++;
+            setMemberVariableValues(word,set_count);
+            break;
+        }
+        if(input[i] ==':'){
+            word = input.substr(startIndex,i-startIndex);
+            startIndex=i+1;
```
Comment (akashshettyjs) [inline TimeParse.cpp:21]: startIndex=i+1; to be done before assigning word

Comment (Ranjeet-8) [reply TimeParse.cpp:21]: if (input[i] == ' ') {
            word = input.substr(startIndex, i - startIndex);
            while(input[++i] == ' ');
            startIndex = i;
            set_count++;
            i--;
            setMemberVariableValues(word, set_count);
        }
        
        
        what about the startIndex in this snippet

File: Display.cpp (hunk 30 above)
Comment (Ranjeet-8) [reply Display.cpp]: done

**Comment by Ranjeet-8** [comment]
code pushed

File: TimeParse.cpp (hunk 31 above)
Comment (akashshettyjs) [reply TimeParse.cpp:21]: here also

Comment (akashshettyjs) [reply TimeParse.cpp:21]: Sorry, Ignore this comment

**Pull request merged**